client/lib/load-script/index.js

```diff
@@ -1,94 +1,65 @@
+/** @format */
 /**
  * A little module for loading a external script
  */
 
 /**
- * Internal dependency.
+ * External dependencies
  */
-var config = require( 'config' );
+import debugFactory from 'debug';
+const debug = debugFactory( 'lib/load-script' );
 
-var JQUERY_URL = 'https://s0.wp.com/wp-includes/js/jquery/jquery.js',
-	callbacksForURLsInProgress = {};
+/**
+ * Internal dependencies
+ */
+import config from 'config';
+import { addScriptCallback, isLoading } from './callback-handler';
+import { createScriptElement, attachToHead } from './dom-operations';
 
-var loadScript = function( url, callback ) {
-	var script = document.createElement( 'script' ),
-		loaded = false;
+// NOTE: This exists for compatibility.
+export { removeScriptCallback } from './callback-handler';
 
-	function handleCompletedRequest( event ) {
-		var errorArgument = null;
-		if ( loaded || ( this.readyState && this.readyState !== 'complete' ) ) {
-			return;
-		}
-		loaded = true;
-		if ( callbacksForURLsInProgress[ url ] ) {
-			if ( event.type === 'error' ) {
-				errorArgument = { src: event.target.src };
-			}
-			callbacksForURLsInProgress[ url ].forEach( function( cb ) {
-				if ( cb ) {
-					cb( errorArgument );
-				}
-			} );
-		}
-		delete callbacksForURLsInProgress[ url ];
-		this.onload = this.onreadystatechange = this.onerror = null;
-	}
+/**
+ * Module variables
+ */
+export const JQUERY_URL = 'https://s0.wp.com/wp-includes/js/jquery/jquery.js';
 
-	// if this url is already being loaded, just add the callback to the queue
-	if ( callbacksForURLsInProgress.hasOwnProperty( url ) ) {
-		callbacksForURLsInProgress[ url ].push( callback );
-		return;
-	}
+//
+// loadScript and loadjQueryDependentScript
+//
+
+export function loadScript( url, callback ) {
+	// If this script is not currently being loaded, create a script element and attach to document head.
+	const shouldLoadScript = ! isLoading( url );
 
-	script.src = url;
-	script.type = 'text/javascript';
-	script.async = true;
-	callbacksForURLsInProgress[ url ] = [ callback ];
-	script.onload = script.onreadystatechange = script.onerror = handleCompletedRequest;
+	addScriptCallback( url, callback );
 
-	document.getElementsByTagName( 'head' )[ 0 ].appendChild( script );
-};
+	if ( shouldLoadScript ) {
+		attachToHead( createScriptElement( url ) );
+	}
+}
```

Comment:
here's a place we can get around some of the DOM-coupling. we can split the export into a raw export and a "default" export

```js
// exported for easy injection of side-effecting browser-coupled actions
export const loadScriptWith = ( { attachToHead, createScriptElement } ) => ( url, callback ) => {
	…
}

// exported with default DOM-manipulating functions
export const loadScript = loadScriptWith( { attachToHead, createScriptElement } );

// actual tests…
import { loadScriptWith } from '../'

const attachToHead = spy();
const createScriptElement = spy();
const loadScript = loadScriptWith( { attachToHead, createScriptElement } );

loadScript( url );
expect( createScriptElement ).to.have.been.calledWith( … )
expect( attachToHead ).to.have.been.calledWith( … )
```

of course you'll not this still leaves the DOM bits untested. they can be tested in isolation if we think we need to test them, but at least at that point we're only testing one thing: can we create a script element; and it won't be attached to the idea of loading specific scripts.

Reply:
I split out DOM operations into a separate file (and corresponding tests). By doing this, I was able to mock out the entire `dom-operations` module in Jest to add tests that check if `createScriptElement` and `attachToHead` functions were called within the `loadScript` function.

Reply:
right, but what I'm suggesting is that we don't need to mock the window functions or load any fake DOM. if all we're doing is creating spies for those functions we can simplify the tests by importing the raw functions and injecting the spies as if they were the window functions. the default export just happens to inject the `window` versions

you don't have to follow this advice, but I think you'll find it easier to mock when we don't have to worry about the actual DOM or `window`

Reply:
I got tired of wrangling the fake DOM approach in testing `dom-operations`, so I decided to go with a hybrid approach: Keep the `dom-operations` module untested, but mock the said module out when testing `load-scripts`. Doing so allows us to mock out `createScriptElement` and `attachToHead` functions without having to deal with the fake DOM provided by jsdom in Jest.